plan.go

```diff
@@ -0,0 +1,172 @@
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// PlanStatus represents a plan state.
+type PlanStatus string
+
+//List all available plan statuses.
+const (
+	PlanCanceled   PlanStatus = "canceled"
+	PlanCreated    PlanStatus = "created"
+	PlanErrored    PlanStatus = "errored"
+	PlanFinished   PlanStatus = "finished"
+	PlanMFAWaiting PlanStatus = "mfa_waiting"
+	PlanPending    PlanStatus = "pending"
+	PlanQueued     PlanStatus = "queued"
+	PlanRunning    PlanStatus = "running"
+)
+
+// Plans handles communication with the plan related methods of the Terraform
+// Enterprise API.
+//
+// TFE API docs: https://www.terraform.io/docs/enterprise/api/plan.html
+type Plans struct {
+	client *Client
+}
+
+// Plan represents a Terraform Enterprise plan.
+type Plan struct {
+	ID               string                `jsonapi:"primary,plans"`
+	HasChanges       bool                  `jsonapi:"attr,has-changes"`
+	LogReadURL       string                `jsonapi:"attr,log-read-url"`
+	Status           PlanStatus            `jsonapi:"attr,status"`
+	StatusTimestamps *PlanStatusTimestamps `jsonapi:"attr,status-timestamps"`
+}
+
+// PlanStatusTimestamps holds the timestamps for individual plan statuses.
+type PlanStatusTimestamps struct {
+	CanceledAt   time.Time `json:"canceled-at"`
+	CreatedAt    time.Time `json:"created-at"`
+	ErroredAt    time.Time `json:"errored-at"`
+	FinishedAt   time.Time `json:"finished-at"`
+	MFAWaitingAt time.Time `json:"mfa_waiting-at"`
+	PendingAt    time.Time `json:"pending-at"`
+	QueuedAt     time.Time `json:"queued-at"`
+	RunningAt    time.Time `json:"running-at"`
+}
+
+// Read a single run by its ID.
+func (s *Plans) Read(ctx context.Context, planID string) (*Plan, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	u := fmt.Sprintf("plans/%s", url.QueryEscape(planID))
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Plan{}
+	err = s.client.do(ctx, req, p)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// Logs retrieves the logs of a plan.
+func (s *Plans) Logs(ctx context.Context, planID string) (io.Reader, error) {
+	if !validStringID(&planID) {
+		return nil, errors.New("Invalid value for plan ID")
+	}
+
+	// Get the plan to make sure it exists.
+	p, err := s.Read(ctx, planID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return an error if the log URL is empty.
+	if p.LogReadURL == "" {
+		return nil, fmt.Errorf("Plan %s does not have a log URL", planID)
+	}
+
+	return &LogReader{
+		client: s.client,
+		ctx:    ctx,
+		planID: p.ID,
+	}, nil
+}
+
+// LogReader implements io.Reader for streaming plan logs.
+type LogReader struct {
+	client *Client
+	ctx    context.Context
+	offset int64
+	planID string
+}
+
+func (r *LogReader) Read(l []byte) (int, error) {
+	for {
+		select {
+		case <-r.ctx.Done():
+			return 0, r.ctx.Err()
+		case <-time.After(500 * time.Millisecond):
```

Comment:
Two things here:

1) This is at the start of the `Read()`, which is a little odd in that it creates an initial "start delay". It seems like this would be better to place after the actual request in case the user wants to do a one-shot `Read()` instead of streaming. What do you think?

2) 500ms is pretty aggressive as a constant delay for streaming - I think the shortest polling interval we have used on the UI side is 1s, and we back off to 3s if no logs are coming in to shed some load. This obviously affects the UX in that users could possibly see more delay, but what do you think about taking a similar approach to guard ourselves from a high volume of API requests?